devito/types/caching.py

```diff
@@ -149,13 +159,21 @@ def clear(cls, force=True):
         sympy.polys.fields._field_cache.clear()
         sympy.polys.domains.modularinteger._modular_integer_cache.clear()
 
+        # Take a copy of the dictionary so we can safely iterate over it
+        # even if another thread is making changes
+
+        # mydict.copy() is safer than list(mydict) for getting an unchanging list
+        # See https://bugs.python.org/issue40327 for terrifying discussion
+        # on this issue.
+        cache_copied = _SymbolCache.copy()
```

Comment:
Isn't it potentially dangerous and may blow up if the cache contains large objects? 

From my understanding, the del/pop part is where we may try to delete already removed object, but it only needs the list of key, not the the full dictionary so  `_SymbolCache.keys()` may be the only thing needed to be copied then access `_SymbolCache[copied_key]`

Reply:
This shouldn’t be too bad I think - as it’s a shallow copy. So won’t duplicate any large objects, just pointers to them. The issue with using keys() is that you have to iterate over it in a non- (less-) gil-safe manner, so we’d need to add more locking. Eg see the iteration over items which failed in CI above. Whereas copy is _mostly_ atomic (notwithstanding some  memory allocation operations which can apparently drop the gil).

Reply:
The memory is fine yes, but because we make a copy, don't we increase the reference count and prevent the actual cache clearing? I guess not since the test pass but seems risky to me, but if all is green I'm probably wrong.

Reply:
Hmmm. I’ll have another look tomorrow to be sure. I _think_ IIRC we’ll just take an extra reference to the weakref object, which won’t keep anything alive that wouldn’t otherwise be.

Reply:
Indeed, the contents of the cache dict are all `AugmentedWeakRef` objects, so I don't think this is an issue. The existence of the copied dict won't change any of the checks in the loop starting at (new) line 186.

Reply:
before we had `list(...)` so wasn't it the same thing practically?

Reply:
Ah, yeah you had `list(x.items())`! So yes, actually we are not copying anything now that we were not copying before.